docs/testing/test-types/README.md

```diff
@@ -0,0 +1,46 @@
+# Test Types
+
+In the LocalStack codebase we differentiate between the following test types:
+
+- Unit tests
+- Acceptance Tests
+- Integration Tests
+
+Depending on the workflow and its trigger not all of those tests are executed at once.
+For ordinary pushes to `master` we only want to execute the Unit and Acceptance tests.
+On a regular schedule, however, we want to execute all tests to have as big of a coverage of our logic as possible.
+This differentiation also educates what we expect from the different types of tests.
+
+## Unit tests
+
+As the name suggests, these tests are performed on smaller units of logic to check if they're sound and perform the operations they claim to.
+This small unit can most often be a kind of helper function inside of a larger procedure.
+These tests should be able to complete their execution very quickly, so they never contain any interaction with some kind of infrastructure.
+If you need some kind of waiting mechanism in your unit test, it is most likely that you are not writing a unit test.
+
+
+## Acceptance tests
+
+We use acceptance tests to gain a quick understanding of whether the recently pushed commit to `master` fulfils minimally viable quality criteria.
+This means that these tests do not aim at maximum coverage but instead should test that the most important functionality works.
+This in general is the entire serving infrastructure and the main features of the most used services.
+
+As these tests are executed very often we need them to be as stable, fast and relevant as possible.
+We ensure this by the following criteria:
+
+- Should perfectly conform to all the testing rules laid out [here](../README.md)
+- Does not contain wait times (e.g., for resources to spin up) averaging longer than 20 seconds
+- Should be able to run in parallel. If tests need to run together (e.g., in a scenario), they need to be added to the same test class.
+- Displays some kind of real-world usage. This can be a scenario with multiple services. 
+    - When composing these scenarios, the services should not overlap too much with already existing acceptance tests. We want to avoid redundancy
+    - Existing samples (from [our samples organization](https://github.com/localstack-samples)) can serve as a good starting point for constructing such a scenario
+- The test should have a proven track record of being stable over a month of being in the regular integration test suite.
+    - Should an acceptance test turn flaky, it will be skipped until it is fixed ([as the rules already state](../README.md))
```

Comment:
I added a few suggested changes to this section:
- I think the 20 seconds were a bit arbitrary, it's hard to define an exact number here. I would just state that they need to be fast.
- At the same time I think that the wording should be strict. For example, the acceptance tests should not be conform to the rules, but they have to.
- I don't think we should introduce a "graduation procedure" for acceptance tests (needs to be stable for x), but we will definitely be super-strict when it comes to skipping flaky ones.

```suggestion
- It is perfectly conform to all the testing rules laid out [here](../README.md)
- It does not contain long wait times (e.g., for resources to spin up). The acceptance tests need to be fast.
- It is fully parallelizable. If certain acceptance tests need to run together (e.g., in a scenario), they need to be added to the same test class.
- It shows some kind of real-world usage. This is usually a scenario with multiple services.
    - When composing these scenarios, the services should not overlap too much with already existing acceptance tests. We want to avoid redundancy where possible. At the same time we want to have our primary services and typical use-cases being covered.
    - Existing samples (from [our samples organization](https://github.com/localstack-samples)) can serve as a good starting point for constructing such a scenario.
- The test needs to be extremely stable.
    - Should an acceptance test turn flaky, it will be skipped until it is fixed ([as we already state in our testing rules](../README.md)).
```

Reply:
- I wanted to be very specific with the execution time. It being "very fast" or waiting times being "long" can mean something else for everyone. The 20 seconds was just a starting point for further discussion. However, we can also decide those things on a case-by-case basis.

- I agree, stricter wording is better here

- I had the concern that adding a lot of acceptance tests of which we don't know the long-term stability yet will defeat the purpose of having a stable set of acceptance tests to rely on. The track record of one month was, like the 20 seconds before, chosen as a starting point for discussion. However, if we accept that there might be some churn in the beginning of adding and skipping tests, we can also remove this requirement.

Reply:
That's totally understandable, but I think we kind of have to decide the execution time requirements on a case-by-case basis because everything else would be a weird guess-game right now. We can distill some educated numbers once we have gained some experience with it.

Concerning the "graduation": I think we should move relatively fast with adding acceptance tests, but be rigorously strict in removing any tests that cause any kind of friction. 😄